mne/preprocessing/stim.py

```diff
@@ -67,3 +79,108 @@ def eliminate_stim_artifact(raw, events, event_id, tmin=-0.005,
+        if inst.preload is False:
+            raise RuntimeError('Modifying data of Raw is only supported '
+                               'when preloading is used. Use preload=True '
+                               '(or string) in the constructor.')
+        events_sel = (events[:, 2] == event_id)
+        event_start = events[events_sel, 0]
+        for k in range(len(event_start)):
+            first_samp = int(event_start[k]) - inst.first_samp + s_start
+            last_samp = int(event_start[k]) - inst.first_samp + s_end
+            data, _ = inst[picks, first_samp:last_samp]
+            if mode == 'linear':
+                x = np.array([first_samp, last_samp])
+                f = interpolate.interp1d(x, data[:, (0, -1)])
+                xnew = np.arange(first_samp, last_samp)
+                interp_data = f(xnew)
+                inst[picks, first_samp:last_samp] = interp_data
+            elif mode == 'window':
+                inst[picks, first_samp:last_samp] = data * window[np.newaxis, :]
+
+    elif isinstance(inst, Epochs):
+        picks = pick_channels(inst.info['ch_names'], inst.info['ch_names'])
+        if inst.preload is False:
+            raise RuntimeError('Modifying data of Epochs is only supported '
+                               'when preloading is used. Use preload=True '
+                               'in the constructor.')
+        if inst.reject:
+            raise RuntimeError('Reject is already applied. Use reject=False '
+                               'in the constructor.')
+        e_start = int(np.ceil(inst.info['sfreq'] * inst.tmin))
+        first_samp = s_start - e_start
+        last_samp = s_end - e_start
+        data = inst.get_data()[:, picks, :]
+        for epoch in data:
+            if mode == 'linear':
+                x = np.array([first_samp, last_samp])
+                f = interpolate.interp1d(x, epoch[:, (0, -1)])
+                xnew = np.arange(first_samp, last_samp)
+                interp_data = f(xnew)
+                epoch[picks, first_samp:last_samp] = interp_data
+            if mode == 'window':
+                epoch[picks, first_samp:last_samp] = \
+                    epoch[picks, first_samp:last_samp] * window[np.newaxis, :]
+        inst._data = data
+
+    elif isinstance(inst, Evoked):
+        picks = pick_channels(inst.info['ch_names'], inst.info['ch_names'])
+        first_samp = s_start - inst.first
+        last_samp = s_end - inst.first
+        data = inst.data
+        if mode == 'window':
+            data[picks, first_samp:last_samp] = \
+                data[picks, first_samp:last_samp] * window[np.newaxis, :]
+        elif mode == 'linear':
+            x = np.array([first_samp, last_samp])
+            f = interpolate.interp1d(x, inst.data[:, (0, -1)])
+            xnew = np.arange(first_samp, last_samp)
+            interp_data = f(xnew)
+            data[picks, first_samp:last_samp] = interp_data
```

Comment:
there is a lot of code duplication with Epochs case. Factorize with a private function.